ClassCastException:  HugeGraphAuthProxy cannot be cast to HugeGraph

### Actual behavior 实际表现
```log
2018-10-23 20:43:00 73074 [main] [ERROR] com.baidu.hugegraph.dist.HugeGraphServer [] - HugeGraphServer error:
java.lang.ClassCastException: com.baidu.hugegraph.auth.HugeGraphAuthProxy cannot be cast to com.baidu.hugegraph.HugeGraph
	at com.baidu.hugegraph.core.GraphManager.checkBackendVersionOrExit(GraphManager.java:172) ~[classes/:?]
	at com.baidu.hugegraph.core.GraphManager.<init>(GraphManager.java:67) ~[classes/:?]
	at com.baidu.hugegraph.server.ApplicationConfig$GraphManagerFactory$1.onEvent(ApplicationConfig.java:102) ~[classes/:?]
	at org.glassfish.jersey.server.internal.monitoring.CompositeApplicationEventListener.onEvent(CompositeApplicationEventListener.java:74) ~[jersey-server-2.25.1.jar:?]
	at org.glassfish.jersey.server.internal.monitoring.MonitoringContainerListener.onStartup(MonitoringContainerListener.java:81) ~[jersey-server-2.25.1.jar:?]
	at org.glassfish.jersey.server.ApplicationHandler.onStartup(ApplicationHandler.java:1180) ~[jersey-server-2.25.1.jar:?]
	at org.glassfish.jersey.grizzly2.httpserver.GrizzlyHttpContainer.start(GrizzlyHttpContainer.java:357) ~[jersey-container-grizzly2-http-2.25.1.jar:?]
	at org.glassfish.grizzly.http.server.HttpHandlerChain.start(HttpHandlerChain.java:422) ~[grizzly-http-server-2.3.28.jar:2.3.28]
	at org.glassfish.grizzly.http.server.HttpServer.setupHttpHandler(HttpServer.java:314) ~[grizzly-http-server-2.3.28.jar:2.3.28]
	at org.glassfish.grizzly.http.server.HttpServer.start(HttpServer.java:290) ~[grizzly-http-server-2.3.28.jar:2.3.28]
	at org.glassfish.jersey.grizzly2.httpserver.GrizzlyHttpServerFactory.createHttpServer(GrizzlyHttpServerFactory.java:296) ~[jersey-container-grizzly2-http-2.25.1.jar:?]
	at org.glassfish.jersey.grizzly2.httpserver.GrizzlyHttpServerFactory.createHttpServer(GrizzlyHttpServerFactory.java:119) ~[jersey-container-grizzly2-http-2.25.1.jar:?]
	at com.baidu.hugegraph.server.RestServer.start(RestServer.java:59) ~[classes/:?]
	at com.baidu.hugegraph.server.RestServer.start(RestServer.java:76) ~[classes/:?]
	at com.baidu.hugegraph.dist.HugeRestServer.start(HugeRestServer.java:54) ~[classes/:?]
	at com.baidu.hugegraph.dist.HugeGraphServer.main(HugeGraphServer.java:58) [classes/:?]
```

### Steps to reproduce the problem 复现步骤
1. enable auth feature
2. start server

### Specifications of environment 环境信息
- hugegraph version: v0.7.4

--- a/hugegraph-api/src/main/java/com/baidu/hugegraph/api/GraphsAPI.java
+++ b/hugegraph-api/src/main/java/com/baidu/hugegraph/api/GraphsAPI.java
@@ -64,7 +64,7 @@ public class GraphsAPI extends API {
     @RolesAllowed({"admin", "$dynamic"})
     public Object list(@Context GraphManager manager,
                        @Context SecurityContext sc) {
-        Set<String> graphs = manager.graphs().keySet();
+        Set<String> graphs = manager.graphs();
         String role = sc.getUserPrincipal().getName();
         if (role.equals("admin")) {
             return ImmutableMap.of("graphs", graphs);

--- a/hugegraph-api/src/main/java/com/baidu/hugegraph/auth/HugeFactoryAuthProxy.java
+++ b/hugegraph-api/src/main/java/com/baidu/hugegraph/auth/HugeFactoryAuthProxy.java
@@ -27,6 +27,9 @@ import com.baidu.hugegraph.auth.HugeGraphAuthProxy.Context;
 
 public class HugeFactoryAuthProxy {
 
+    public static final String GRAPH_FACTORY =
+           "gremlin.graph=com.baidu.hugegraph.auth.HugeFactoryAuthProxy";
+
     static {
         HugeGraphAuthProxy.setContext(Context.admin());
     }

--- a/hugegraph-api/src/main/java/com/baidu/hugegraph/auth/StandardAuthenticator.java
+++ b/hugegraph-api/src/main/java/com/baidu/hugegraph/auth/StandardAuthenticator.java
@@ -82,6 +82,9 @@ public class StandardAuthenticator extends SimpleAuthenticator {
     @Override
     public AuthenticatedUser authenticate(final Map<String, String> credentials)
                                           throws AuthenticationException {
+        if (!this.requireAuthentication()) {
+            return AuthenticatedUser.ANONYMOUS_USER;
+        }
         String username = credentials.get(PROPERTY_USERNAME);
         String password = credentials.get(PROPERTY_PASSWORD);
 

--- a/hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java
+++ b/hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java
@@ -19,6 +19,7 @@
 
 package com.baidu.hugegraph.core;
 
+import java.util.Collections;
 import java.util.HashSet;
 import java.util.Map;
 import java.util.Set;
@@ -34,6 +35,7 @@ import org.apache.tinkerpop.gremlin.structure.util.GraphFactory;
 import org.slf4j.Logger;
 
 import com.baidu.hugegraph.HugeGraph;
+import com.baidu.hugegraph.auth.HugeFactoryAuthProxy;
 import com.baidu.hugegraph.auth.HugeGraphAuthProxy;
 import com.baidu.hugegraph.auth.StandardAuthenticator;
 import com.baidu.hugegraph.backend.cache.Cache;
@@ -68,21 +70,19 @@ public final class GraphManager {
     }
 
     public void loadGraphs(final Map<String, String> graphConfs) {
-        graphConfs.entrySet().forEach(conf -> {
+        for (Map.Entry<String, String> conf : graphConfs.entrySet()) {
+            String name = conf.getKey();
+            String path = conf.getValue();
             try {
-                final Graph newGraph = GraphFactory.open(conf.getValue());
-                this.graphs.put(conf.getKey(), newGraph);
-                LOG.info("Graph '{}' was successfully configured via '{}'",
-                         conf.getKey(), conf.getValue());
+                this.loadGraph(name, path);
             } catch (RuntimeException e) {
-                LOG.error("Graph '{}': '{}' can't be instantiated",
-                          conf.getKey(), conf.getValue(), e);
+                LOG.error("Graph '{}' can't be loaded: '{}'", name, path, e);
             }
-        });
+        }
     }
 
-    public Map<String, Graph> graphs() {
-        return this.graphs;
+    public Set<String> graphs() {
+        return Collections.unmodifiableSet(this.graphs.keySet());
     }
 
     public HugeGraph graph(String name) {
@@ -106,8 +106,7 @@ public final class GraphManager {
     }
 
     public void rollbackAll() {
-        this.graphs.entrySet().forEach(e -> {
-            final Graph graph = e.getValue();
+        this.graphs.values().forEach(graph -> {
             if (graph.features().graph().supportsTransactions() &&
                 graph.tx().isOpen()) {
                 graph.tx().rollback();
@@ -120,8 +119,7 @@ public final class GraphManager {
     }
 
     public void commitAll() {
-        this.graphs.entrySet().forEach(e -> {
-            final Graph graph = e.getValue();
+        this.graphs.values().forEach(graph -> {
             if (graph.features().graph().supportsTransactions() &&
                 graph.tx().isOpen()) {
                 graph.tx().commit();
@@ -137,9 +135,9 @@ public final class GraphManager {
                          final Transaction.Status tx) {
         final Set<Graph> graphsToCloseTxOn = new HashSet<>();
 
-        graphSourceNamesToCloseTxOn.forEach(r -> {
-            if (this.graphs.containsKey(r)) {
-                graphsToCloseTxOn.add(this.graphs.get(r));
+        graphSourceNamesToCloseTxOn.forEach(name -> {
+            if (this.graphs.containsKey(name)) {
+                graphsToCloseTxOn.add(this.graphs.get(name));
             }
         });
 
@@ -163,9 +161,21 @@ public final class GraphManager {
         return this.authenticator.authenticate(username, password);
     }
 
+    private void loadGraph(String name, String path) {
+        final Graph graph = GraphFactory.open(path);
+        this.graphs.put(name, graph);
+        LOG.info("Graph '{}' was successfully configured via '{}'", name, path);
+
+        if (this.authenticator.requireAuthentication() &&
+            !(graph instanceof HugeGraphAuthProxy)) {
+            LOG.warn("You may need to support access control for '{}' with {}",
+                     path, HugeFactoryAuthProxy.GRAPH_FACTORY);
+        }
+    }
+
     private void checkBackendVersionOrExit() {
-        for (Graph graph : this.graphs.values()) {
-            HugeGraph hugegraph = (HugeGraph) graph;
+        for (String graph : this.graphs()) {
+            HugeGraph hugegraph = this.graph(graph);
             if (InMemoryDBStoreProvider.matchType(hugegraph.backend())) {
                 continue;
             }
@@ -176,6 +186,7 @@ public final class GraphManager {
                 System.exit(-1);
             }
             if (!info.checkVersion()) {
+                // Exit if versions are inconsistent
                 System.exit(-1);
             }
         }
@@ -183,7 +194,7 @@ public final class GraphManager {
 
     private void addMetrics(HugeConfig config) {
         final MetricManager metrics = MetricManager.INSTANCE;
-        // Force add server reporter
+        // Force to add server reporter
         ServerReporter reporter = ServerReporter.instance(metrics.getRegistry());
         reporter.start(60L, TimeUnit.SECONDS);
 
@@ -220,8 +231,8 @@ public final class GraphManager {
             String exp = String.format("%s.%s", key, "expire");
             String size = String.format("%s.%s", key, "size");
             String cap = String.format("%s.%s", key, "capacity");
-            // Avoid register multi times
 
+            // Avoid registering multiple times
             if (names.stream().anyMatch(name -> name.endsWith(hits))) {
                 continue;
             }

